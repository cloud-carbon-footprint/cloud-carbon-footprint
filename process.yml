# Orb 'circleci/node@4.0.0' resolved to 'circleci/node@4.0.0'
# Orb 'circleci/gcp-gcr@0.8.0' resolved to 'circleci/gcp-gcr@0.8.0'
# Orb 'circleci/gcp-cli@1.8.4' resolved to 'circleci/gcp-cli@1.8.4'
version: 2
jobs:
  deploy:
    docker:
    - image: hashicorp/terraform:0.12.28
    steps:
    - checkout
    - run:
        name: install gcloud
        command: |
          apk add --update python3 py-crcmod bash libc6-compat curl
          export PATH=$PATH:/root/google-cloud-sdk/bin
          curl -sSL https://sdk.cloud.google.com > /tmp/gcloud-install && bash /tmp/gcloud-install --disable-prompts --install-dir=/root
    - run:
        command: |
          IFS="," read -ra PARAMS <<< "GCLOUD_SERVICE_KEY,GOOGLE_PROJECT_ID,GOOGLE_COMPUTE_ZONE"

          for i in "${PARAMS[@]}"; do
            if [[ -z "${i}" ]]; then
              echo "ERROR: Missing environment variable {i}" >&2

              if [[ -n "" ]]; then
                echo "" >&2
              fi

              exit 1
            else
              echo "Yes, ${i} is defined!"
            fi
          done
        name: Checking whether required env vars are set to initialize gcloud CLI...
    - run:
        command: |
          # Store service account
          echo $GCLOUD_SERVICE_KEY > ${HOME}/gcloud-service-key.json

          # Initialize gcloud CLI
          gcloud auth activate-service-account --key-file=${HOME}/gcloud-service-key.json
          gcloud --quiet config set project $GOOGLE_PROJECT_ID
          gcloud --quiet config set compute/zone $GOOGLE_COMPUTE_ZONE
        name: Initialize gcloud CLI to connect to Google Cloud
workflows:
  version: 2
  build-and-deploy:
    jobs:
    - deploy

# Original config.yml file:
# version: 2.1
# orbs:
#   node: circleci/node@4.0.0
#   gcr: circleci/gcp-gcr@0.8.0
#   gcp-cli: circleci/gcp-cli@1.8.4
# jobs:
#   build:
#     working_directory: ~/cloud-carbon-footprint
#     docker:
#       - image: circleci/node:current
#     environment:
#       MAJOR_REVISION: 0
#       MINOR_REVISION: 0
#       GIT_HASH: <<pipeline.git.revision>>
#     # Currently, the cache is commented out since it's slower than installing dependencies every time. Once the
#     # dependency list grows and build time increases, we can start using the cache again. Caching takes about ~30s to
#     # resolve and use as of this comment.
#     steps:
#       - checkout
#       # - restore_cache:
#       #     key: dependency-cache-{{ checksum \"package-lock.json\" }}-{{ checksum \"client/package-lock.json\" }}-{{ checksum \"server/package-lock.json\" }}
#       - run:
#           name: setup
#           command: \"[ ! -d node_modules ] && npm run bootstrap || echo NPM packages unchanged. Using cache.\"
#       # - save_cache:
#       #     key: dependency-cache-{{ checksum \"package-lock.json\" }}-{{ checksum \"client/package-lock.json\" }}-{{ checksum \"server/package-lock.json\" }}
#       #     paths:
#       #       - ./node_modules
#       #       - ./client/node_modules
#       #       - ./server/node_modules
#       - run:
#           name: lint
#           command: 'npm run lint'
#       - run:
#           name: test
#           command: 'npm test'
#   publish-client:
#     description: Build and push client image to Google Container Registry
#     machine:
#       image: ubuntu-1604:202007-01
#     steps:
#       - checkout
#       - run:
#           name: build
#           command: |
#             cd client
#             npm install
#             npm run-script build
#       - gcr/gcr-auth
#       - gcr/build-image:
#           docker-context: client
#           path:  client
#           image: client
#           tag: 0.0.<<pipeline.git.revision>>
#       - gcr/push-image:
#           image: client
#           tag: 0.0.<<pipeline.git.revision>>
#   publish-server:
#     description: Build and push server image to Google Container Registry
#     machine:
#       image: ubuntu-1604:202007-01
#     steps:
#       - checkout
#       - run:
#           name: build
#           command: |
#             cd server
#             npm install
#             npm run-script build
#       - gcr/gcr-auth
#       - gcr/build-image:
#           docker-context: server
#           path:  server
#           image: server
#           tag: 0.0.<<pipeline.git.revision>>
#       - gcr/push-image:
#           image: server
#           tag: 0.0.<<pipeline.git.revision>>
#   deploy:
#     description: Deploy app to gcp
#     docker:
#       - image: hashicorp/terraform:0.12.28
#     steps:
#       - checkout
#       - run: 
#           name: install gcloud
#           command: |
#             apk add --update python3 py-crcmod bash libc6-compat curl
#             export PATH=$PATH:/root/google-cloud-sdk/bin
#             curl -sSL https://sdk.cloud.google.com > /tmp/gcloud-install && bash /tmp/gcloud-install --disable-prompts --install-dir=/root
#       - gcp-cli/initialize
#       # - run:
#       #     environment:
#       #       TF_VAR_server_version: 0.0.<<pipeline.git.revision>>
#       #       TF_VAR_client_version: 0.0.<<pipeline.git.revision>>
#       #     name: deploy
#       #     command: |
# 
#       #       cd terraform
#       #       ./run.sh apply
# workflows:
#   version: 2
#   build-and-deploy:
#     jobs:
#       # - build
#       # - publish-client:
#       #     requires:
#       #       - build
#       # - publish-server:
#       #     requires:
#       #       - build
#       - deploy
#           # requires:
#           #   - publish-client
#           #   - publish-server