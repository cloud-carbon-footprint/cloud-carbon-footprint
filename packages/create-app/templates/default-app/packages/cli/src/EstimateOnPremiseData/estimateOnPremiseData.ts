/*
 * Â© 2021 Thoughtworks, Inc.
 */
import process from 'process'
import commander from 'commander'
import csv from 'csvtojson'
import path from 'path'

import { App } from '@cloud-carbon-footprint/app'
import {
  OnPremiseDataInput,
  OnPremiseDataOutput,
} from '@cloud-carbon-footprint/common'
import { validateOnPremiseInput } from '../common/inputValidation'
import { writeToCsv } from './writeToCsv'

export default async function estimateOnPremiseData(
  argv: string[] = process.argv,
): Promise<void> {
  const program = new commander.Command()
  program.storeOptionsAsProperties(false)

  program
    .option(
      '--onPremiseInput <filename>',
      'File containing relevant On-Premise input data',
    )
    .option(
      '--onPremiseOutput <filename>',
      'File to write the relevant On-Premise output data',
      'on_premise_estimations.csv',
    )

  program.parse(argv)

  const programOptions = program.opts()

  const onPremiseInputFile = programOptions.onPremiseInput
  const onPremiseOutputFile = path.join(
    process.cwd(),
    programOptions.onPremiseOutput,
  )

  if (onPremiseInputFile) {
    const onPremiseInputData: OnPremiseDataInput[] = await csv().fromFile(
      onPremiseInputFile,
    )
    validateOnPremiseInput(onPremiseInputData)
    const onPremiseEstimatesData: OnPremiseDataOutput[] =
      new App().getOnPremiseEstimatesFromInputData(onPremiseInputData)
    await writeToCsv(onPremiseOutputFile, onPremiseEstimatesData)
  }
}
