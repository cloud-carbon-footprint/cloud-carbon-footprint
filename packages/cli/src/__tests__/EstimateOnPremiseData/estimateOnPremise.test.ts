/*
 * Â© 2021 Thoughtworks, Inc.
 */

import fs from 'fs'
import path from 'path'

import estimateOnPremiseData from '../../EstimateOnPremiseData/estimateOnPremiseData'
import process from 'process'

describe('estimateOnPremiseData', () => {
  let inputFilePath: string
  let outputFilePath: string

  describe('creates on premise estimates CSV', () => {
    afterEach(() => {
      try {
        fs.unlinkSync(outputFilePath)
      } catch (err) {
        console.log(err)
      }
    })
    it('creates on premise CSV file, with default output file path', async () => {
      inputFilePath = path.join(__dirname, 'on_premise_data_input.test.csv')
      outputFilePath = path.join(process.cwd(), 'on_premise_estimations.csv')

      await estimateOnPremiseData([
        'executable',
        'file',
        '--onPremiseInput',
        inputFilePath,
      ])

      expect(fs.existsSync(outputFilePath)).toBe(true)
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot()
    })

    it('creates on premise CSV file, with provided output file name', async () => {
      inputFilePath = path.join(__dirname, 'on_premise_data_input.test.csv')
      outputFilePath = path.join(
        process.cwd(),
        'on_premise_estimations_data.csv',
      )

      await estimateOnPremiseData([
        'executable',
        'file',
        '--onPremiseInput',
        inputFilePath,
        '--onPremiseOutput',
        'on_premise_estimations_data.csv',
      ])

      expect(fs.existsSync(outputFilePath)).toBe(true)
      expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot()
    })
  })

  it('Throws if the input data file is incorrect', async () => {
    inputFilePath = path.join(__dirname, 'incorrect_input.test.csv')

    await expect(() =>
      estimateOnPremiseData([
        'executable',
        'file',
        '--onPremiseInput',
        inputFilePath,
      ]),
    ).rejects.toThrow(
      'Input data is incorrect. Please check your input data file and try again.',
    )
  })
})
