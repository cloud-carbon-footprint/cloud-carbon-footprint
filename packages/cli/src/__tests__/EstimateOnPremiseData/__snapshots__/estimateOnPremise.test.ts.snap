// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`estimateOnPremiseData creates on premise estimates CSV creates on premise CSV file, with default output file path 1`] = `
"cpuId,memory,machineType,startTime,endTime,country,region,cpuUtilization,powerUsageEffectiveness,usageHours,kilowattHours,co2e
Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz,130690,server,2022-01-17T13:38:18Z,2022-01-24T18:22:29.918423Z,United States,Texas,,,172,98.95676,0.0409997648032
"
`;

exports[`estimateOnPremiseData creates on premise estimates CSV creates on premise CSV file, with provided output file name 1`] = `
"cpuId,memory,machineType,startTime,endTime,country,region,cpuUtilization,powerUsageEffectiveness,usageHours,kilowattHours,co2e
Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz,130690,server,2022-01-17T13:38:18Z,2022-01-24T18:22:29.918423Z,United States,Texas,,,172,98.95676,0.0409997648032
"
`;
